server/src/main/java/io/druid/timeline/partition/HashBasedNumberedShardSpec.java

```diff
@@ -24,28 +24,43 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.base.Function;
 import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
 import com.google.common.hash.HashFunction;
 import com.google.common.hash.Hashing;
 import io.druid.data.input.InputRow;
 import io.druid.data.input.Rows;
 
+import javax.annotation.Nullable;
 import java.util.List;
 
 public class HashBasedNumberedShardSpec extends NumberedShardSpec
 {
   private static final HashFunction hashFunction = Hashing.murmur3_32();
+  private static final List<String> DEFAULT_PARTITION_DIMENSIONS = ImmutableList.of();
+
   private final ObjectMapper jsonMapper;
+  private final List<String> partitionDimensions;
 
   @JsonCreator
   public HashBasedNumberedShardSpec(
       @JsonProperty("partitionNum") int partitionNum,
       @JsonProperty("partitions") int partitions,
+      @JsonProperty("partitionDimensions") @Nullable List<String> partitionDimensions,
```

Comment:
For realtime ingestion, people only use none/linear/numbered shardSpec, and HashBasedNumberedShardSpec and SingleDimensionShardSpec are used internal.
http://druid.io/docs/latest/ingestion/realtime-ingestion.html